runtest.csh uses POSIX launcher

rhwood opened the pull request:
Make runtest.csh a wrapper around an ant process that will ensure the test class is run from a launcher built from the current AppScriptTemplate for POSIX launchers.

rhwood commented:
This works by using ant to generate a script ```run.sh``` that ```runtest.csh``` then calls. I'm not sure if ```run.sh``` should be included in git and regenerated using ```ant run-sh``` as needed, or added to .gitignore to require that it always be (re)created before use.

rhwood commented:
I realized that by renaming ```run.sh``` to ```.run.sh```, tab completion to call runtest.csh works better, and excluding .run.sh from git is less likely to cause confusion than excluding run.sh.

pabender commented:
What's the time penalty incurred by running the full wrapper script?

Part of the reason we haven't done this previously is because of the added time involved when running individual tests.

rhwood commented:
> What's the time penalty incurred by running the full wrapper script?

Running .run.sh directly adds about 0.1 seconds over the unmodified runtest.csh. Having ant verify that .run.sh is current adds about 2/3 of a second, or slightly longer if the ".run.sh" script needs to be (re)generated.

Note that if you want to run .run.sh directly with the effects of having been wrapped in runtest.csh, you would need to run (from the project root directory): ```mkdir ./temp ; cp ./local.conf ./temp/jmri.conf ; ./.run.sh --settingsdir=./temp CLASS ARGS...```

> Part of the reason we haven't done this previously is because of the added time involved when running individual tests.

You leave a lot at the table by doing do, like the classpath appending abilities in #1472, and the ability to set memory limits, neither of which cannot be affected by adding a java property to the ```JMRI_OPTIONS``` environment; it appears these kinds of capabilities are desired for this from comments in that issue.

If running tests against java code, the fastest way to run them is ```ant test-single CLASS``` unless you need to pass arguments to the test; since with runtest.csh, you have to ```ant tests ; runtest.csh CLASS```

I'm not sure what you gain by skipping ensuring that the tests are actually compiled every time within runtest.csh, since when I asked earlier, I received no reply.

pabender commented:
On Jul 23, 2016, at 7:05 AM, Randall Wood <notifications@github.com> wrote:
> What's the time penalty incurred by running the full wrapper script?
> 
> Running .run.sh directly adds about 0.1 seconds over the unmodified runtest.csh. Having ant verify that .run.sh is current adds about 2/3 of a second, or slightly longer if the ".run.sh" script needs to be (re)generated.
> 
That's a lot of overhead when you're running a single test class that takes under .1s to run,
> Note that if you want to run .run.sh directly with the effects of having been wrapped in runtest.csh, you would need to run (from the project root directory): mkdir ./temp ; cp ./local.conf ./temp/jmri.conf ; ./.run.sh --settingsdir=./temp CLASS ARGS...
> 
If the test needs to interact with the configuration directories, that should be configured by the testing infrastructure using either an @Before or @BeforeClass method and cleaned up using either an @After or @AfterClass method. 

By using a single configuration directory here, you contaminate the test environment for the second test using that directory.
> Part of the reason we haven't done this previously is because of the added time involved when running individual tests.
> 
> You leave a lot at the table by doing do, like the classpath appending abilities in #1472, and the ability to set memory limits, neither of which cannot be affected by adding a java property to the JMRI_OPTIONS environment; it appears these kinds of capabilities are desired for this from comments in that issue.
> 
But there isn't any reason for any of that when running package or class tests.  Unless the test is specifically testing for interaction with the environment, running the tests should be self contained, and independent of the environmental settings.   If they aren't, then there is something wrong with the test.
> If running tests against java code, the fastest way to run them is ant test-single CLASS unless you need to pass arguments to the test; since with runtest.csh, you have to ant tests ; runtest.csh CLASS
> 
> I'm not sure what you gain by skipping ensuring that the tests are actually compiled every time within runtest.csh, since when I asked earlier, I received no reply.
> 
One reason is I want to verify the result of a test before I actually run updated code.  In other words, the workflow is:

1) Make changes to code
2) run tests with runtest.csh
3) if tests errors, make corrections, which may drive the error off the screen
4) re-run tests to get the errors back
5) repeat steps 4 and 5 until finished with update
6) rebuild code and tests
7) repeat steps 2-6 as necessary.

So I specifically don't want the tests rebuilt before I explicitly ask for that in step 6.  That allows me to verify that I have addressed the test issue(s) completely before I run the new test.

Paul

rhwood commented:
> By using a single configuration directory here, you contaminate the test environment for the second test using that directory.

You specifically stated this was a good idea in #1597.

pabender commented:
On Jul 23, 2016, at 10:43 AM, Randall Wood <notifications@github.com> wrote:
> > By using a single configuration directory here, you contaminate the test environment for the second test using that directory.
> 
> You specifically stated this was a good idea in #1597.

In #1597, I was referring to files and directories created by the tests.  The change you made here creates the directory before the tests are run AND puts a file in that directory.  Setting up the external environment is very much a task that should be handled by the @Before or @BeforeClass methods, not by a script that runs around them.

Paul

bobjacobsen commented:
We discussed the utility of `./runtest.csh` before: Some people care about the speed and simplicity of it.

When I repeat the test from #1584, I see that this is a regression on the primary purpose of `runtest.csh`:

````
% time ./runtest.csh jmri.BundleTest
......
Time: 0.021

OK (6 tests)

2.011u 0.246s 0:02.01 111.9%	0+0k 0+10io 0pf+0w
````

Further, I think it's now broken. If I set optional parameters via JMRI_OPTIONS, it no longer works:

````
JMRI/projects/JMRI/ % setenv JMRI_OPTIONS -Djmri.skipschematests=true
JMRI/projects/JMRI/ % time ./runtest.csh jmri.BundleTest
JUnit version 4.12
.E
Time: 0.003
There was 1 failure:
1) initializationError(org.junit.runner.JUnitCommandLineParseResult)
java.lang.IllegalArgumentException: Could not find class [-Djmri.skipschematests=true]
	at org.junit.runner.JUnitCommandLineParseResult.parseParameters(JUnitCommandLineParseResult.java:102)
	at org.junit.runner.JUnitCommandLineParseResult.parseArgs(JUnitCommandLineParseResult.java:50)
	at org.junit.runner.JUnitCommandLineParseResult.parse(JUnitCommandLineParseResult.java:44)
	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:72)
	at org.junit.runner.JUnitCore.main(JUnitCore.java:36)
	at jmri.util.junit.TestClassMainMethod.main(TestClassMainMethod.java:25)
Caused by: java.lang.ClassNotFoundException: -Djmri.skipschematests=true
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.junit.internal.Classes.getClass(Classes.java:16)
	at org.junit.runner.JUnitCommandLineParseResult.parseParameters(JUnitCommandLineParseResult.java:100)
	... 5 more

FAILURES!!!
Tests run: 1,  Failures: 1

2.098u 0.243s 0:02.02 115.3%	0+0k 0+4io 0pf+0w
````
True, there are other mechanisms, but this one is still supposed to work (and I have test automation that still uses it)

Finally, what's the hurry in all this?  Why was this merged with known, outstanding concerns?  Is it fixing some user-level problem?

rhwood commented:
The is the final item that resolves #1472 on POSIX platforms, it prevents the issues in #1597, as well as prevents JMRI tests from picking up configuration from ```settings:```.

The ```-D...``` arguments being passed as args to ```main()``` is an error in the POSIX launcher in general; I'm working on a separate PR for that (I tested using ```-J...``` arguments), so arguably, this is a test of the POSIX launchers in general.

pabender commented:
Sent from my iPad

On Jul 23, 2016, at 4:09 PM, Bob Jacobsen <notifications@github.com> wrote:

> We discussed the utility of ./runtest.csh before: Some people care about the speed and simplicity of it.
> 
> 
<snip>
> Finally, what's the hurry in all this? Why was this merged with known, outstanding concerns? Is it fixing some user-level problem?
> 

I agree with Bob on this.  This was merged without addressing the concerns of the scripts users.

I still have two pretty major concerns about this:
1) the script invokes ant, with all of ants overhead.  That breaks the speed and simplicity of the 
runtest.csh script.

2) The script adds a jmri.conf file to the temp directory containing the settings.  If jmri.conf is 
required for a test, that test should create jmri.conf in either the @Before or @BeforeClass 
method.  This is actually the bigger concern for me because that means runtest.csh is broken 
for ANY test that tests loading of options using the jmri.conf mechanisim.

Until these concerns are addressed, I think this change should be reverted.

Paul

rhwood commented:
> On Jul 23, 2016, at 4:09 PM, Bob Jacobsen <notifications@github.com> wrote:
> 
> > We discussed the utility of ./runtest.csh before: Some people care about the speed and simplicity of it.
> > 
> > 
> <snip>
> > Finally, what's the hurry in all this? Why was this merged with known, outstanding concerns? Is it fixing some user-level problem?

I was reading two issues I’ve filed with very specific suggestions for fixing those issues and this fixes those issues as I proposed and it appears the commenters on this PR have approved of.

Bob explicitly asked that the fixes for modifying how classpath are handled in #1472 be "the same thing for all the various ways we launch” and this helps implement that request. Paul specifically stated that “would be a good idea” in response to #1597 to "to ensure that ant and any processes that run tests ensure that the jmri.prefsdir property is set to a temporary file (or to a directory ignored by git within a local repository) before tests are run”.

> I agree with Bob on this. This was merged without addressing the concerns of the scripts users.
> 
> I still have two pretty major concerns about this:
> 1) the script invokes ant, with all of ants overhead. That breaks the speed and simplicity of the 
> runtest.csh script.

If you look in master, you will see that it only invokes ant after performing bash tests to see if invoking ant really is needed so as to avoid that overhead.

> 2) The script adds a jmri.conf file to the temp directory containing the settings. If jmri.conf is 
> required for a test, that test should create jmri.conf in either the @Before or @BeforeClass 
> method. This is actually the bigger concern for me because that means runtest.csh is broken 
> for ANY test that tests loading of options using the jmri.conf mechanism.

jmri.conf by design provides a persistent mechanism indistinguishable from setting JMRI_OPTIONS except that its in a file instead of the environment, the purpose of which affect the JVM at or prior to launch (such as setting the max heap size or appending or prepending classes to the classpath). If having that file really is an issue, then its an issue no different than allowing JMRI_OPTIONS to be set at all, in which case runtest.csh should explicitly ensure that JMRI_OPTIONS is unset. However it is clear that the expectation is that the environment in which runtest.csh is used can be adjusted, so there is no clear argument that storing the equivalent of JMRI_OPTIONS in a file is any worse than setting JMRI_OPTIONS in the environment.

In master, runtest.csh uses a fixed jmri.conf file, instead of copying one around, so this is a non-issue.

pabender commented:
On 07/23/2016 08:54 PM, Randall Wood wrote:
>> On Jul 23, 2016, at 4:09 PM, Bob Jacobsen <notifications@github.com>
> wrote:
>>
>> > We discussed the utility of ./runtest.csh before: Some people care
> about the speed and simplicity of it.
>> >
>> >
>> <snip>
>> > Finally, what's the hurry in all this? Why was this merged with
> known, outstanding concerns? Is it fixing some user-level problem?
>
> I was reading two issues I’ve filed with very specific suggestions for
> fixing those issues and this fixes those issues as I proposed and it
> appears the commenters on this PR have approved of.
>
> Bob explicitly asked that the fixes for modifying how classpath are
> handled in #1472 be "the same thing for all the various ways we launch”
> and this helps implement that request. Paul specifically stated that
> “would be a good idea” in response to #1597 to "to ensure that ant and
> any processes that run tests ensure that the jmri.prefsdir property is
> set to a temporary file (or to a directory ignored by git within a local
> repository) before tests are run”.

I'm not complaining about setting the jmri.prefsdir property.  Though 
that CAN be handled in the TestClassMainMethod class.


>> I agree with Bob on this. This was merged without addressing the
> concerns of the scripts users.
>>
>> I still have two pretty major concerns about this:
>> 1) the script invokes ant, with all of ants overhead. That breaks the
> speed and simplicity of the
>> runtest.csh script.
>
> If you look in master, you will see that it only invokes ant after
> performing bash tests to see if invoking ant really is needed so as to
> avoid that overhead.

So it means that any time I run `ant clean`, ant is invoked.

>> 2) The script adds a jmri.conf file to the temp directory containing
> the settings. If jmri.conf is
>> required for a test, that test should create jmri.conf in either the
> @Before or @BeforeClass
>> method. This is actually the bigger concern for me because that means
> runtest.csh is broken
>> for ANY test that tests loading of options using the jmri.conf mechanism.
>
> jmri.conf by design provides a persistent mechanism indistinguishable
> from setting JMRI_OPTIONS except that its in a file instead of the
> environment, the purpose of which affect the JVM at or prior to launch
> (such as setting the max heap size or appending or prepending classes to
> the classpath). If having that file really is an issue, then its an
> issue no different than allowing JMRI_OPTIONS to be set at all, in which
> case runtest.csh should explicitly ensure that JMRI_OPTIONS is unset.
> However it is clear that the expectation is that the environment in
> which runtest.csh is used can be adjusted, so there is no clear argument
> that storing the equivalent of JMRI_OPTIONS in a file is any worse than
> setting JMRI_OPTIONS in the environment.

It's possible for non-JVM JMRI_OPTIONS to be over-ridden by an 
individual test or a group of tests by over-riding the methods that get 
the system variables and/or by using a custom JUnit4 Runner class.

Our test coverage is pretty poor in this area, and this is something we 
need to work on, so I doubt this is handled correctly now.

There are several suggestions for how to handle this type of thing in 
this stackoverflow thread:

http://stackoverflow.com/questions/8168884/how-to-test-code-dependent-on-environment-variables-using-junit

These all amount to wrapping the environment value retrieval in another 
class so that you can mock the actual value you want.

I think it's actually possible to test the JVM related

> In master, runtest.csh uses a fixed jmri.conf file, instead of copying
> one around, so this is a non-issue.

My question is how do we test the mechanism that loads jmri.conf.  As 
long as that file is constant, We can't do that.

Paul
